src/lib/utils/thorchain/hooks/useSendThorTx.tsx

```diff
@@ -0,0 +1,426 @@
+import { ExternalLinkIcon } from '@chakra-ui/icons'
+import { Link, Text, useToast } from '@chakra-ui/react'
+import type { AccountId, AssetId } from '@shapeshiftoss/caip'
+import { fromAccountId, fromAssetId, thorchainAssetId } from '@shapeshiftoss/caip'
+import type { FeeDataEstimate } from '@shapeshiftoss/chain-adapters'
+import { CONTRACT_INTERACTION, FeeDataKey } from '@shapeshiftoss/chain-adapters'
+import { SwapperName } from '@shapeshiftoss/swapper'
+import type { KnownChainIds } from '@shapeshiftoss/types'
+import { useQuery } from '@tanstack/react-query'
+import dayjs from 'dayjs'
+import { useCallback, useMemo, useState } from 'react'
+import { useTranslate } from 'react-polyglot'
+import { reactQueries } from 'react-queries'
+import { selectInboundAddressData } from 'react-queries/selectors'
+import { getAddress, zeroAddress } from 'viem'
+import type { SendInput } from 'components/Modals/Send/Form'
+import { estimateFees, handleSend } from 'components/Modals/Send/utils'
+import { useWallet } from 'hooks/useWallet/useWallet'
+import { bn, bnOrZero } from 'lib/bignumber/bignumber'
+import { getTxLink } from 'lib/getTxLink'
+import { fromBaseUnit, toBaseUnit } from 'lib/math'
+import { assertUnreachable, isToken } from 'lib/utils'
+import { assertGetThorchainChainAdapter } from 'lib/utils/cosmosSdk'
+import {
+  assertGetEvmChainAdapter,
+  buildAndBroadcast,
+  createBuildCustomTxInput,
+} from 'lib/utils/evm'
+import { THORCHAIN_POOL_MODULE_ADDRESS } from 'lib/utils/thorchain/constants'
+import { depositWithExpiry } from 'lib/utils/thorchain/routerCalldata'
+import { useGetEstimatedFeesQuery } from 'pages/Lending/hooks/useGetEstimatedFeesQuery'
+import { THORCHAIN_SAVERS_DUST_THRESHOLDS_CRYPTO_BASE_UNIT } from 'state/slices/opportunitiesSlice/resolvers/thorchainsavers/utils'
+import {
+  selectAccountNumberByAccountId,
+  selectAssetById,
+  selectFeeAssetByChainId,
+  selectSelectedCurrency,
+} from 'state/slices/selectors'
+import { serializeTxIndex } from 'state/slices/txHistorySlice/utils'
+import { useAppSelector } from 'state/store'
+
+import { fromThorBaseUnit, getThorchainTransactionType } from '..'
+
+type Action =
+  | 'swap'
+  | 'addLiquidity'
+  | 'withdrawLiquidity'
+  | 'openLoan'
+  | 'repayLoan'
+  | 'depositSavers'
+  | 'withdrawSavers'
+
+type UseSendThorTxProps = {
+  accountId: AccountId | null
+  action: Action
+  amountCryptoBaseUnit: string | undefined
+  assetId: AssetId | undefined
+  enableEstimateFees?: boolean
+  disableEstimateFeesRefetch?: boolean
+  fromAddress: string | null
+  memo: string | undefined
```

Comment:
can we unify the use of undefined or null in this type?

Reply:
Consolidated to `null`: 7d43b0b6f7. Though things sometimes get a bit messy with opt chains which inherently return `undefined`. Perhaps we should bring back the `Nullable` type i.e `T | null | undefined`?

Reply:
I would really prefer to see `undefined` over `null`. Considering it is the default nullish value, I don't see a good reason to ever use `null` unless we are handling it being directly returned from a dependency package imo